hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/SnapshotProcedure.java

```diff
@@ -0,0 +1,422 @@
+      }
+    }
+  }
+
+  @Override
+  protected boolean isRollbackSupported(SnapshotState state) {
+    return true;
+  }
+
+  @Override
+  protected SnapshotState getState(final int stateId) {
+    return SnapshotState.forNumber(stateId);
+  }
+
+  @Override
+  protected int getStateId(SnapshotState state) {
+    return state.getNumber();
+  }
+
+  @Override
+  protected SnapshotState getInitialState() {
+    return SnapshotState.SNAPSHOT_PREPARE;
+  }
+
+  private void prepareSnapshot(MasterProcedureEnv env)
+    throws ProcedureSuspendedException, IOException {
+    if (isAnySplitOrMergeProcedureRunning(env)) {
+      if (retryCounter == null) {
+        retryCounter = ProcedureUtil.createRetryCounter(env.getMasterConfiguration());
+      }
+      long backoff = retryCounter.getBackoffTimeAndIncrementAttempts();
+      LOG.warn("{} waits {} ms for Split/Merge procedure to finish", this, backoff);
+      setTimeout(Math.toIntExact(backoff));
+      setState(ProcedureState.WAITING_TIMEOUT);
+      skipPersistence();
+      throw new ProcedureSuspendedException();
+    }
+    prepareSnapshotEnv(env);
+  }
+
+  private void prepareSnapshotEnv(MasterProcedureEnv env) throws IOException {
+    this.conf = env.getMasterConfiguration();
+    this.snapshotTable = TableName.valueOf(snapshot.getTable());
+    this.htd = loadTableDescriptorSnapshot(env);
+    this.rootDir = CommonFSUtils.getRootDir(conf);
+    this.snapshotDir = SnapshotDescriptionUtils.getCompletedSnapshotDir(snapshot, rootDir);
+    this.workingDir = SnapshotDescriptionUtils.getWorkingSnapshotDir(snapshot, rootDir, conf);
+    this.workingDirFS = workingDir.getFileSystem(conf);
+    this.status = TaskMonitor.get()
+      .createStatus("Taking " + snapshot.getType() + " snapshot on table: " + snapshotTable);
+    ForeignExceptionDispatcher monitor = new ForeignExceptionDispatcher(snapshot.getName());
+    this.snapshotManifest = SnapshotManifest.create(conf,
+      workingDirFS, workingDir, snapshot, monitor, status);
+    this.snapshotManifest.addTableDescriptor(htd);
+  }
+
+  @Override
+  protected synchronized boolean setTimeoutFailure(MasterProcedureEnv env) {
+    setState(ProcedureState.RUNNABLE);
+    env.getProcedureScheduler().addFront(this);
+    return false;
+  }
+
+  private boolean isAnySplitOrMergeProcedureRunning(MasterProcedureEnv env) {
+    return env.getMasterServices().getMasterProcedureExecutor().getProcedures().stream()
+      .filter(p -> !p.isFinished())
+      .filter(p -> p instanceof SplitTableRegionProcedure ||
+        p instanceof MergeTableRegionsProcedure)
+      .anyMatch(p -> ((AbstractStateMachineTableProcedure<?>) p)
+        .getTableName().equals(getTableName()));
+  }
+
+  private TableDescriptor loadTableDescriptorSnapshot(MasterProcedureEnv env) throws IOException {
+    TableDescriptor htd = env.getMasterServices().getTableDescriptors().get(snapshotTable);
+    if (htd == null) {
```

Comment:
Is this possible?

Reply:
It exists in the old code. I can't think of a situation where this would happen. Actually I tried removing this check and it seemed that it didn't make any difference. The unit tests still passed. This is just to keep the same as the old code.

Reply:
Mind giving a pointer in the current code base? We could check the history.

Reply:
OK. It is in the file 

https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/master/snapshot/TakeSnapshotHandler.java